core/server/worker/src/main/java/alluxio/worker/netty/DataServerWriteHandler.java

```diff
@@ -170,129 +255,202 @@ private boolean tooManyPacketsInFlight() {
               mPosToQueue);
     }
-    return "";
-  }
-
-  /**
-   * Writes an error block write response to the channel and closes the channel after that.
-   *
-   * @param channel the channel
-   */
-  private void replyError(Channel channel, Protocol.Status.Code code, String message, Throwable e) {
-    channel.writeAndFlush(RPCProtoMessage.createResponse(code, message, e, null))
-        .addListener(ChannelFutureListener.CLOSE);
-  }
-
-  /**
-   * Writes a response to signify the success of the block write. Also resets the channel.
-   *
-   * @param channel the channel
-   */
-  private void replySuccess(Channel channel) {
-    channel.writeAndFlush(RPCProtoMessage.createOkResponse(null))
-        .addListeners(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE, new ChannelFutureListener() {
-          @Override
-          public void operationComplete(ChannelFuture future) throws Exception {
-            reset();
-          }
-        });
-    if (!channel.config().isAutoRead()) {
-      channel.config().setAutoRead(true);
-      channel.read();
-    }
-  }
-
-  private void reset() throws IOException {
-    if (mRequest != null) {
-      mRequest.close();
-      mRequest = null;
-    }
-
-    mPosToQueue = 0;
-    mPosToWrite = 0;
-
-    try {
-      mLock.lock();
-      for (ByteBuf buf : mPackets) {
-        ReferenceCountUtil.release(buf);
-      }
-      mPacketWriterActive = false;
-    } finally {
-      mLock.unlock();
+    if (msg.getPayloadDataBuffer() != null && msg.getPayloadDataBuffer().getLength() > 0 && (
+        request.getCancel() || request.getEof())) {
+      return String.format("Found data in a cancel/eof message.");
     }
+    return "";
   }
 
   /**
    * A runnable that polls from the packets queue and writes to the block worker.
    */
   private final class PacketWriter implements Runnable {
-    private ChannelHandlerContext mCtx;
+    private Channel mChannel;
 
     /**
      * Creates an instance of {@link PacketWriter}.
      *
-     * @param ctx the netty channel handler context
+     * @param channel the netty channel
      */
-    PacketWriter(ChannelHandlerContext ctx) {
-      mCtx = ctx;
+    PacketWriter(Channel channel) {
+      mChannel = channel;
     }
 
     @Override
     public void run() {
       try {
         runInternal();
       } catch (Throwable e) {
+        // This should never happen.
         LOG.error("Failed to run PacketWriter.", e);
         throw e;
       }
     }
 
-    /**
-     * The actual implementation of the runnable.
-     */
     private void runInternal() {
+      boolean eof;
+      boolean cancel;
+      boolean abort;
+
       while (true) {
         ByteBuf buf;
         mLock.lock();
         try {
           buf = mPackets.poll();
-          if (buf == null) {
+          if (buf == null || buf == EOF || buf == CANCEL || buf == ABORT) {
+            eof = buf == EOF;
+            cancel = buf == CANCEL;
+            abort = (buf == ABORT || mError != null);
```

Comment:
mError overrides ABORT. ABORT is strictly not necessary. I have added a TODO. I feel it is cleaner to follow what i did in DataServerReadHandler.